[FLINK-4186] Use Flink metrics to report Kafka metrics

rmetzger opened the pull request:
We were using Flink's accumulators in the past to report the Kafka metrics.
With this change, we'll use the new metrics of Flink.

I'm now also exposing the current offset of each topic partition through the metrics.

rmetzger commented:
This will simplify this PR: https://issues.apache.org/jira/browse/FLINK-4206

zentol commented:
You can rebase on top of #2237.

zentol commented on flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java:
```diff
@@ -161,6 +167,12 @@ public void runFetchLoop() throws Exception {
 			}
 
+			// register offset metrics
+			if(useMetrics) {
```
missing space after if

zentol commented on flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java:
```diff
@@ -78,7 +79,10 @@
+	private final long autoCommitInterval;
+
+	/** The metric group of this operator */
+	private final MetricGroup metricGroup;
```
the 0.8 Fetcher has a `MetricGroup` field, while the 0.9 Fetcher only creates a local variable inside run.  Let's make this more consistent.

zentol commented on flink-streaming-connectors/flink-connector-kafka-0.8/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/Kafka08Fetcher.java:
```diff
@@ -161,6 +167,12 @@ public void runFetchLoop() throws Exception {
 
+			// register offset metrics
+			if(useMetrics) {
+				final MetricGroup kafkaMetricGroup = metricGroup.addGroup("Kafka08Consumer");
```
i don't think it is necessary to include the version in the group name.

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/AbstractFetcher.java:
```diff
@@ -389,8 +396,36 @@ private void updateMinPunctuatedWatermark(Watermark nextWatermark) {
 	}
-	
-	// ------------------------------------------------------------------------
+
```
double empty line

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/metrics/KafkaMetricGetter.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.flink.metrics.Gauge;
+import org.apache.kafka.common.Metric;
+
+
```
double empty line, also missing javadocs.

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/testutils/MockRuntimeContext.java:
```diff
@@ -161,6 +163,11 @@ public Histogram getHistogram(String name) {
 
 	@Override
+	public MetricGroup getMetricGroup() {
+		return new UnregisteredTaskMetricsGroup.DummyIOMetricGroup();
```
You don't really have to override this method, using the TaskMetricGroup is fine.

rmetzger replied:
I was getting a null pointer exception, that's why I return dummy metrics here.

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1235,15 +1235,104 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
+	}
+
+	/**
+	 * Test metrics reporting for consumer
```
Can we have a test for the producer as well?

rmetzger commented:
Thank you for the review @zentol.
I addressed all your comments and added metrics for committed offsets as well (I talked to the user who requested this feature and they requested this additional metric).

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumerBase.java:
```diff
@@ -65,6 +65,9 @@
 	public static final int MAX_NUM_PENDING_CHECKPOINTS = 100;
 
+	/** Boolean configuration key to disable metrics tracking **/
+	public static final String KEY_DISABLE_METRICS = "flink.disable-metrics";
```
is this key set in `flink-conf.yaml` ?

rmetzger replied:
No, its passed as a property into the consumers.

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/internals/metrics/KafkaMetricGetter.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Gauge for getting the current value of a Kafka metric.
+ */
+public class KafkaMetricGetter implements Gauge<Double> {
```
calling this KafkaMetricWrapper would be more in line with the current naming scheme for these things.

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
+						numPosOffsets++;
+					}
+				}
+				if(numPosOffsets == 5) {
```
missing space after if

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
+			// we can't rely on the consumer to have touched all the partitions already
+			// that's why we'll wait until all five partitions have a positive offset.
+			// The test will fail if we never meet the condition
+			while(true) {
```
space after while

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
+			JobManagerCommunicationUtils.cancelCurrentJob(flink.getLeaderGateway(timeout));
+		}
+
+		while(jobThread.isAlive()) {
```
space after while

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1235,15 +1236,129 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
+		while(jobThread.isAlive()) {
+			Thread.sleep(50);
+		}
+		if(error.f0 != null) {
```
space after if

rmetzger commented:
I fixed the failing build and addressed all comments so far ;) (I was working on the code while you've reviewed it )

zentol commented on flink-streaming-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaConsumerTestBase.java:
```diff
@@ -1235,15 +1236,127 @@ public void flatMap(Tuple2<Integer, Integer> value, Collector<Void> out) throws
+				// check that offsets are correctly reported
+				for (ObjectName object : offsetMetrics) {
+					Object offset = mBeanServer.getAttribute(object, "Value");
+					if((long) offset >= 0) {
```
missing space

zentol commented:
couldn't find any problem, +1 from my side.

rmetzger commented:
I'm merging the change ...
